### PR TITLE
[SDESK-4448] Administrator can manage private templates

### PR DESCRIPTION
**Pull request by marwoodandrew**

### CODE REVIEW

**Comment by tomaskikutis** [comment]
FYI you don't need travis to pass.

We have many unstable e2e tests and are working around this with retries in fireq. We started running e2e test without retries on travis and are logging the unstable ones for fixing later so we can eventually drop fireq when the tests are stabilized.

**Comment by marwoodandrew** [comment]
@tomaskikutis Thanks

File: scripts/apps/templates/directives/TemplatesDirective.ts [hunk 1]
```diff
@@ -348,6 +364,10 @@ export function TemplatesDirective(notify, api, templates, modal, desks, weekday
                 {label: gettext('No Desk'), value: 'None'},
             ];
 
+            if (templates.isAdmin(true)) {
+                $scope.filters.splice(2, 0, {label: gettext('Private'), value: 'Private'});
```
Comment (tomaskikutis) [inline scripts/apps/templates/directives/TemplatesDirective.ts:368]: Can you use `push` instead? It's easier to follow the logic when it's sequential without having to wonder what makes `2` the special position.

File: scripts/apps/templates/services/TemplatesService.ts [hunk 2]
```diff
@@ -112,9 +113,11 @@ export function TemplatesService(api, session, $q, preferencesService, privilege
 
         return $q.when(criteria)
             .then((criteriaReady) => {
-                params.where = JSON.stringify({
-                    $and: [criteriaReady],
-                });
+                if (!_.isEmpty(criteriaReady)) {
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply scripts/apps/templates/services/TemplatesService.ts:116]: Checking for an empty object, I think this is a reasonable way to do it and is consistent with other similar existing checks in the same source file.

[the remark this answers is not in the document]
Comment (tomaskikutis) [reply scripts/apps/templates/services/TemplatesService.ts:116]: What I don't like about `isEmpty` is that it operates on many types: objects, arrays, buffers, strings. On the one hand, it's convenient, but on the other, it makes it harder to understand what data structures are being manipulated. Can you then add types for `criteriaReady`?

File: scripts/apps/templates/directives/TemplatesDirective.ts (hunk 1 above)
Comment (marwoodandrew) [reply scripts/apps/templates/directives/TemplatesDirective.ts:372]: Done

File: scripts/apps/templates/filters/index.ts [hunk 3]
```diff
@@ -7,30 +7,37 @@ import _ from 'lodash';
...
-                return !item.is_public;
+                return !item.is_public && (session.identity._id === item.user);
+            case 'Private':
+                return !item.is_public && (session.identity._id !== item.user);
             default:
                 return _.find(item.template_desks, (desk) => desk === f.value);
             }
         });
+
+        if (f.value === 'Private') {
```
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply scripts/apps/templates/filters/index.ts]: Done

File: scripts/apps/templates/directives/TemplatesDirective.ts (hunk 1 above)
Comment (tomaskikutis) [reply scripts/apps/templates/directives/TemplatesDirective.ts:372]: The splice is still here. Can you use `push` for `{label: gettext('Private'), value: 'Private'}` too?

Comment (marwoodandrew) [reply scripts/apps/templates/directives/TemplatesDirective.ts:372]: oops that should have gone! Private gets pushed conditionally on line 367

File: scripts/apps/templates/services/TemplatesService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (marwoodandrew) [reply scripts/apps/templates/services/TemplatesService.ts:116]: What about using 
if (Object.keys(criteriaReady).length)

[the remark this answers is not in the document]
Comment (tomaskikutis) [reply scripts/apps/templates/services/TemplatesService.ts:116]: :+1: that's exactly what I meant by an explicit check. You might add a default value handling if null/undefined are expected `if (Object.keys(criteriaReady ?? {}).length > 0)`

**Pull request merged**